File: 0003.LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.c
int lengthOfLongestSubstring(char *s)
{
        char *allc[256] = {0};
        char *left, *right;
        int maxlen = 0;
        int slen = strlen(s);

        left = right = s;
        while (*right)
        {
                if (allc[*right] != NULL && allc[*right] - left >= 0)
                {
                        maxlen = right - left > maxlen ? right - left : maxlen;
                        left = allc[*right] + 1;
                }
                allc[*right] = right;
                right++;
        }

        return right - left > maxlen ? right - left : maxlen;
}

File: 0003.LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        x = {}
        start = 0
        end = 0
        strlen = len(s)
        maxlen = 0

        for end in range(strlen):
            if s[end] in x and x[s[end]] >= start:
                maxlen = max(end - start, maxlen)
                start = x[s[end]] + 1
            x[s[end]] = end
        maxlen = max(strlen - start, maxlen)

        return maxlen
